### ISSUE TITLE
Enhancement: Search Component UI Control 

### ISSUE THREAD

**Issue by thabaum** [oqtane/oqtane.framework#3291]
Enhancement request for a Search Component UI Control.  

I already did the work! 2 days ;) as I started it yesterday and love my results.  I had to stop myself from going any further as I want to ask about a "Button" UI Control as well I may create off what I created if desired by the community.

I know there will be lots of opinions on this but bare with me please.  I am ready to adjust and let @sbwalker adjust as desired.

It is fairly complex... it passes a lot of customizable parameters and does everything but bring you coffee!

It will be able to search from any other parent component and is localized.

Did I say it has options?  Lots of options... all I could think of in the past 24 hours.

Some eye candy:

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/fdf4ffab-1c15-4f38-af24-6c7e25440a69)

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/36ad80eb-b419-40d5-925a-c217fd8bf96d)

You can add remove the button, add remove the icon, put the button left or right side of input, put the icon right or left side of button text... Search-as-you-type, enter key enable/disable, no button, localization, accessibility, plus more... and it is a completely reusable component passing parameters and data fields to search as desired...

My only problem is my VS is not connected to GitHub and I put this up the hard way (editing each file online right from the source repo)

I may have to actually get the VS working on Github this time as I am editing the Page Management file to show its use, a Search.razor component, a Search.resx file and of coarse a little taste of some app.css flavoring.

I am very pleased with the results but I am sure it can be fine tuned better.

Here is an example use of it from settings on screenshot and I dont have all parameters here you can pass as a lot are defaulted in but you can change virtually everything possible almost with ease as desired for type of search functionality you would like to have in a component that uses this:

```
    <ActionLink Action="Add" Text="Add Page" ResourceKey="AddPage" />
    <Search DataSource="allResults"
            IsSearchEnabled="true"
            MinCharacters="0"
            InputValidationRegex="@SearchLocalizer["SearchValidationRegex"].Value"
            Placeholder="@SearchLocalizer["SearchPlaceholder"].Value"
            InputAriaLabel="@SearchLocalizer["InputAriaLabel"].Value"
            ButtonAriaLabel="@SearchLocalizer["ButtonAriaLabel"].Value"
            ButtonToolTip="@SearchLocalizer["ButtonToolTip"].Value"
            ShowSearchButton="true"
            EnterKeyEnabled="true"
            ResultsAsTyped="false"
            DebounceEnabled="false"
            CaseSensitive="false"
            @bind-searchTerm="searchTerm"
            ItemTypes='new List<string> { "Name" }'
            InputTabIndex = "0"
            ButtonTabIndex = "1"
            OnSearchResultsChanged="HandleSearchResults" />
    <Pager Items="@pageList">
```

Playing with all these other UI Libraries I am liking the idea of making an Oqtane UI Library as complete as any other out there.

Fun stuff eh?

I will submit a PR for review by end of the day.  I hope it satisfies the Oqtane Community and I look forward to all the feedback.

So before creating one take my component for a spin once I can upload this PR.

**Comment by sbwalker** [comment]
This looks interesting... this seems to be a natural fit as part of the Pager component (which is essentially how you implemented it in the screen shots) - are there any other scenarios where it would be used? It seems like a quick win to integrate it with the Pager to allow all of the Admin pages to benefit from search capabilities. 

In regards to UI components, the philosophy behind Oqtane is to keep the core framework as small and lean as possible (for performance reasons). Originally I did not want to include any UI components at all. However based on the fact that Oqtane includes an Admin UI, I concluded that it would be necessary to include a limited number of UI components to help with maintainability and consistency for the admin experience. But to be clear, the main goal from a UI perspective is avoid creating UI components for Oqtane - and to provide the integration capability so that developers can use their preferred UI component libraries.

**Comment by thabaum** [comment]
I put a long version of reply in the PR to help explain how I believe this should align with Oqtane's "No Third Party" and "Developer Friendly"  aspects.

I would really love to avoid third parties.  And I did think of making a UI library module, in fact maybe it can be another project for Oqtane such as the Blog.

But I thought so many out there already, it would be nice not to have to chase licensing requirements for those of us not looking to be restricted by them.  Oqtane licensing, granted it never changes, aligns with a free for all for any purpose in any way use case.  No other UI library is like this as they all go "Pro" version eventually.  And when you install the libraries you add more dependencies potentially.

Oqtane UI Controls are no dependent on any third party, just HTML more less with some C#/Blazor mixed with it to function.

This UI control has no third party added as I used everything already included in Oqtane.

Pretty much if you want to use Oqtane and start using third parties you got a job ahead of you rewriting all the admin pages to work with the new UI libraries, along with the other sources of support you need to angle.

Like the recently added Icons feature, UI Controls should be in my view added to Oqtane and enhanced to a desirable level to get most common tasks done similar to the Blazor QuickGrid.  UI Controls are being added to .NET Core itself.  It will be included in .NET 8 for production use.

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/quickgrid?view=aspnetcore-7.0](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/quickgrid?view=aspnetcore-7.0)

Maybe a QuickChart, QuickSearch, QuickButton down the road who knows as also options coming later.  Maybe you know something I don't in respect to this which is why UI controls are not of interest in the framework?

**Comment by sbwalker** [comment]
@thabaum you touched on a point above which is a fairly widespread misconception when it comes to Blazor development...

Based on the fact that Blazor has a component model, it leads people to believe that everything should be abstracted as an individual component. I have seen this referred to as "micro-componentization" where applications are broken down to the most primitive UI components, and then those components are used within "macro" components, which are then used within application components, etc... This all makes sense in theory based on object oriented principles.

However there is a real performance cost to using so many components, as each component has its own state, events, and its own life cycle. The rendering time for deeply nested component render trees is very expensive. I would encourage every Blazor developer to read this guidance from Microsoft:

https://learn.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-7.0#avoid-thousands-of-component-instances

And I realize that most of the popular UI component libraries provide micro-components for every possible UI element... and .NET developers like this approach (likely because it is "familiar" to them based on prior desktop or mobile development experiences)... but there are real performance impacts (I have personally helped clients diagnose Blazor performance issues which were caused by UI component libraries).

The overarching theme for Oqtane is "performance" - therefore Oqtane has tried to avoid micro-components in its Admin UI and has utilized standard HTML markup as much as possible. I believe this provides more context as to why I do not want to introduce primitives like a "button" component. And also why I asked if it would make sense to integrate this into the Pager component rather than include it as a stand-alone component.

Another useful piece of Blazor guidance is related to component parameters:

https://learn.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-7.0#dont-receive-too-many-parameters

So although it seems like a good idea to include a lot of parameters for customization reasons, fewer parameters is better for performance reasons.

**Comment by thabaum** [comment]
I agree performance is an issue, it is great to see there is some literature to read relating as I always thought there would be.  At the same time I figure CPUs and Memory are going to make this less an issue for example Intels new CPU.  This along with all the performance increases with each version of .NET and the fact like you said all the mainstream UI libraries have these.

It's really such a tiny bit of information in most UI controls, there are default values so you dont need to pass anything unless you want to modify it past the original settings.

After reading that article... which you are right about the misconception here and thank you for bringing this more to my attention.

As far as basic UI components

#Avoid thousands of component instances
This includes things like buttons, but if you just have one like in the search button it should work ok eh?
#Don't receive too many parameters
Again with the defaults set, you really just need to pass the data, the more feature settings you want to adjust will create a little more overhead for the CPU/Memory/Cache to render and maintain.

I guess my OOP instincts are proven to be an issue here.

So the use of these UI libraries potentially bring more than just a security risk due to the additional potential surface attacks so to speak like Telerik introduced into DNN, but the performance issues using them in components also brings a negative affect.

I agree then this could be included in the Pager itself.  I have no desire now to use any UI libraries, but instead create everything per component.

You would think that adding a component to a component when the app is built, the components join hands and remove these performance issues in some form.

Does or should the Pager potentially use Virtualization?

I did use render fragments for the button creation...

However this is another concern with this UI search control and others like it I can also notice to take into consideration:

[https://learn.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-7.0#avoid-attribute-splatting-with-captureunmatchedvalues](https://learn.microsoft.com/en-us/aspnet/core/blazor/performance?view=aspnetcore-7.0#avoid-attribute-splatting-with-captureunmatchedvalues)

On the other hand, this search UI control is very small, has one button, one input... not a lot of components given you dont use a button UI control as well which even then it is only 2 components, not thousands.

Using a button component in the Pager however would be devastating to performance without virtualization possibly if I understand correctly?

**Comment by sbwalker** [comment]
The Pager does not need virtualization because it only displays a fixed number of rows from the dataset - not the entire data set.

**Comment by thabaum** [comment]
Let me get a better grip on virtualization here.  When the data set is retrieved it loads the entire data set into memory correct?  Virtualization would only retrieve the 10 to be displayed and not get the entire data set?

I may not fully understand this but the pager gets the entire data set but then displays 10 of them.  When I used the search it was instantly getting the page with the as you type feature enabled.

I am trying to completely understand the use case for Virtualization as it seems anything connecting to data with say 100,000 pages for example, I take it this been tested to only need to get the first 10/20/50/100?  Then when you hit page 2 it gets the next 10 so it never loads the entire data set.  Is this correct if I am clear I hope?

Thing with the pager is you set it to 100 and that is all you get, the next 100,000 errors in a log file do not get loaded limiting the pager to 100 pages potentially?

Seems like I felt that is how it was treating data sets, but I maybe wrong.

**Comment by sbwalker** [comment]
@thabaum no you are misunderstanding the Virtualize component. It's purpose is solely for rendering efficiency - it has nothing to do with how the data is loaded into a component. The Virtualize component renders only the items that are currently visible on the page  ie. the portion visible on the screen.

**Comment by sbwalker** [comment]
@thabaum once I get a chance to review your PR in more detail I will provide some feedback on the preferred approach for integration into the framework.

**Comment by thabaum** [comment]
It sounds like a green light for adding this at least to the pager itself, I would not mind seeing it enhanced it further with sorting options similar to .NET 8 QuickGrid if that is OK?

Would you like to use the code in the PR or I can submit a new one with the logic from the Search control merged into the pager?  Or you can if you feel you can provide the options you would be interested in including?

It was worth the learning experience regardless of who or how search in pager can be added.

If you like I can attempt the PR as you can see how I was implementing this, but you probably will have a better strategy to accomplish the results Oqtane can use.

I will close my other PR.

**Comment by sbwalker** [comment]
@thabaum I said that once I get a chance to review your PR in more detail I will provide some feedback on the preferred approach for integration into the framework (I have not had time to review it yet - too many other PRs and big fixes today)

**Comment by sbwalker** [comment]
@thabaum I had some time to review your PR this morning so I will provide my feedback below...

- Oqtane already contains 2 components for filtering lists of data - the AutoComplete component and InputList component. These components have distinct purposes and are integrated into the Admin UI in specific areas.
- your Search component appears to be very similar to the AutoComplete component which is used in the permission grid to allow an administrator to select a user based on a portion of a user's name. The AutoComplete component was designed to be used in any scenario where you have a reasonable amount of data and need "type ahead" or "auto complete" filtering functionality.
- it would not make sense to have 2 components which essentially do the same thing- however there may be aspects of your search component which could be incorporated in the AutoComplete component. This would need further analysis.
- from a technical perspective your Search component has debounce logic using a timer. Timers create memory leaks unless they are properly disposed. Your component does not have dispose logic.

- if we focus on the use case of an enhancement to the Pager component which is used throughout the Admin UI - it would be useful to allow users to filter the list of data to locate specific items more efficiently
- the Pager search should be pre-configured with defaults so that for the majority of cases a developer would not need to specify any extra parameters for the Pager component. 
- the number of parameters related to search functionality should be kept to a minimum - more parameters equals more complexity, more maintenance, and more testing permutations
- The one parameters which I am conflicted about is a ShowSearch parameter. It would potentially be a breaking UI behavioral change to enable search by default (ie. if an existing user of the Pager has already implemented their own search functionality) so the default may need to be False and a developer would need to explicitly set it to True to enable the functionality.
- the Pager search could utilize the data already loaded into the client-side component by default. A parameter would likely be needed to indicate which columns in the list should be targeted.  
- Optionally there could also be an event which could be implemented which could all a method in the parent component if the developer wanted to implement  server-side search logic (not required).
- The Pager search should also have a Reset button which restores the data to the original unfiltered results (instead of forcing them to delete the search text and select Search).
- the search input and buttons should use the input-group class to group UI elements to be consistent with the rest of the Admin UI

**Comment by thabaum** [comment]
@sbwalker thanks for taking time to review.  

I did see the logic in the AutoComplete component and noticed it was close, but my thinking was a different use case.  Since autocomplete filters data similar it should show some similarities.

This entire PR was really to help give some ideas.

Did you test without button or icon ResultsAsTyped=True, EnterKeyEnabled=False, ShowButtonSearchButton=false, debounce=false.

And I must of missed that timer memory leak as I had something to deal with it but I may have removed it along all the testing trials I made.

Integration with pager issue, maybe the resx file can have the ShowSearch or change to EnableSearch setting for true/false to enable/disable?

Maybe the autocomplete can be included into this, I was thinking about doing it but still just wanted some initial feedback on how this first attempt would go issue wise to see what can be resolved to be acceptable.

We can use input-group class as I wasnt sure exactly how to add the inputs and buttons into Oqtane Framework considering so many different ways to do it anymore.

And really the defaults should all be there for the framework to only really need to send the data source and item type to filter using currently in this PR "ItemTypes='new List<string> { "Name" }'" where you can add all the columns you would like to add for the filter to search.

The resetting of data was something of concern.  They ResultsAsTyped works great if you delete what you have, but there could also be an option to turn on a reset button.  I was trying to make something that would get options to do whatever for more than just the pager, but  I am for whatever works best for the framework.

The search button as soon as you hit enter the field is cleared, hit enter again and you get all results.  I figured this area is where the "fine tuning" would come into place to get desirable results.

If you have defaults set, then you dont need to pass parameters you can limit it.  But if you want to customize it I added those parameters to do so if needed for special use case examples.  

For using with the pager I was trying to keep it as minimal or complex a configuration as a developer using it would need.  

I wanted to add the AutoComplete feature and was hoping to hear something back on that.  So the feedback is close to what I expected.

Bottom line as a  search component this should be the minimal you need to pass with all other defaults in place:

```
    <ActionLink Action="Add" Text="Add Page" ResourceKey="AddPage" />
    <Search DataSource="allResults"
            @bind-searchTerm="searchTerm"
            ItemTypes='new List<string> { "Name" }'
            OnSearchResultsChanged="HandleSearchResults" />
    <Pager Items="@pageList">

```

**Comment by thabaum** [comment]
Thanks again for all the feedback, just to clear anything up I can resolve most issues but we just need to come up with all the final thoughts as to how this should fit into the Oqtane Framework for developers.  

Pushing all these UI components into an Oqtane.NativeUIComponents module including all previous ones would be a bonus as it would allow the complete removal of all of these if not needed easier.  Something down the road maybe?

•	Oqtane already contains 2 components for filtering lists of data - the AutoComplete component and InputList component. These components have distinct purposes and are integrated into the Admin UI in specific areas.
o	The “distinct” purpose is why I did not merge this PR into those areas, AutoComplete specifically as I did not review InputList and will now I am more aware of its functionality.
•	your Search component appears to be very similar to the AutoComplete component which is used in the permission grid to allow an administrator to select a user based on a portion of a user's name. The AutoComplete component was designed to be used in any scenario where you have a reasonable amount of data and need "type ahead" or "auto complete" filtering functionality.
o	This has AutoComplete but I have not tested it as one of the settings that comes stock in input fields.
•	it would not make sense to have 2 components which essentially do the same thing- however there may be aspects of your search component which could be incorporated in the AutoComplete component. This would need further analysis.
o	Could be the other way around too, with AutoComplete being added into the Search.  I was hoping to maybe use it as an option to enable which would only bring the child component total to 2-3.
•	from a technical perspective your Search component has debounce logic using a timer. Timers create memory leaks unless they are properly disposed. Your component does not have dispose logic.
o	I was trying all the different methods to produce debounce logic.  I have to say this was my last piece I was adding to help with keeping the searching calls down but was not very satisfied with end results.  I was hoping this could be fine tuned and maybe there was a better solution to use than Timer in .NET core.  I did see it used in the framework so I went for it.  I also had memory leak disposal but I may have removed it while copying/pasting logic from all my other pages of test code.  So this I was hoping to have a deeper discussion or see it fine tuned how it should be to function in a desirable way when enabled.  This was actually one of my last enhancements and I really wanted some guidance to see if this Timer method is the correct one to use for this use case and .NET Core Blazor WebAssembly.  I felt like there maybe other options I am not aware of or something in .NET 8 that was better to use.  I did want to add the feature here to review so it was included as an enhancement to discuss.

Pager integration
•	it would be useful to allow users to filter the list of data in the Pager to locate specific items more efficiently
o	can you elaborate more on “more efficiently” or is this an OK we should add this feature to the pager?
•	the Pager search should be pre-configured with defaults so that for the majority of cases a developer would not need to specify any extra parameters for the Pager component.
o	I agree which is why I was hoping you would come up with the best configuration to be presented as default tailored as Oqtane Admin defaults.
•	the number of parameters related to search functionality should be kept to a minimum - more parameters equals more complexity, more maintenance, and more testing permutations
o	I agree here, I just don’t know if having parameters to send is the same issue as actually sending them.  The defaults should help minimize these concerns by choosing only the ones that are most important to change from default for the purpose used by a developers use case.
•	the one parameter which I am conflicted about is a ShowSearch parameter. It would potentially be a breaking UI behavioral change to enable search by default (ie. if an existing user of the Pager has already implemented their own search functionality) so the default may need to be False and a developer would need to explicitly set it to True to enable the functionality.
o	My thoughts here is maybe this can be set in the search resource file?
•	the Pager search could utilize the data already loaded into the client-side component by default. A parameter would likely be needed to indicate which columns in the list should be targeted.
o	This is already there, just maybe poorly named.  ItemTypes is a list “Name” “Summary” “Description” of column(s) to be searched.
•	Optionally there could also be an event which could be implemented which could call a method in the parent component if the developer wanted to implement server-side search logic (not required).
o	I would need a better understanding here of how this would be used since the UI controls are created on pages with API calls to server-side, client-side I may not fully understand this option, but any option that may be usable I would love to get in place so it can be used if this is added to Oqtane for all use cases a developer may need if possible.
•	The Pager search should also have a Reset button which restores the data to the original unfiltered results (instead of forcing them to delete the search text and select Search).
o	I did think about how to add this as well and if it was needed in all developer use cases.  Potential solution is options for using an close X in the input field/ option for next to button right/left with/without icon.
•	the search input and buttons should use the input-group class to group UI elements to be consistent with the rest of the Admin UI
o	I wasn’t 100% sure which input elements and classes to use, this can surely be adjusted and should be.  The “app-search” class wrapper should do the rest as needed for further customization of styling (pin stripes, racing stripes, spoiler, ground effects, …)

Off to the races for the weekend, check back later after Sunday for further feedback and give you more time to review how this should be properly implemented to enhance the framework.

**Comment by thabaum** [comment]
I had some time today to address most issues here:

- input-group
- debouncing working as expected and disposal as needed when component is no longer in use.
- Clear button and function
- naming conventions to C#/Oqtane best practices
- Tested additional parameters such as "Name", "Path" and even without the results of the Path field type if I typed mypage all the pages with mypage including children showed up in results as expected.  Didnt test it but figured it should work since Name worked.
- Added error messaging below input-group along with more defense logic  (still have some reviewing here)

To Do:
- finish localization efforts
- finish defaults
- create less parameters (I was going to create a case statement for different types of search behaviors for the 4 I have introduced to take 4 available parameters down to 1)
- Review all accessibility options are properly applied
- review and set any desirable options on how the clear button reacts to the search button locations
- Potentially having a parameter to set that if more than one type of field of the data source is searched the fields are selectable dropdown for searching (ie. ALL - Pages - Path - IconFile) such as EnableTypeSelection = false for default but if you have more than one it will create a dropdown list if desired next to the search button, or the clear button, or.... wherever... in the input-group.

I know one issue is not to use <object> but use <Page> for example due to performance issues having to store more from what I read in memory.  So maybe there is a work around where a parameter can be passed inside where <object> goes?  But I think this is wishful thinking...

I will submit updates tomorrow once I get all of these accomplished as the hard stuff is complete I believe.

For disabled/enabled we can have a setting in the resource file that is disabled default but you can enable for "PagerSearchEnabled" set to false for upgrades, true for new installations.

@sbwalker I will be getting this all done tomorrow I think if I can get time after a few jobs I have to handle are done. if not I will shoot for tuesday so you can see the updates, review code and test functionalities.

I will just need to learn about how to make the Oqtane Framework upgrade/install logic for the two different scenarios.  That will be a final review after all other issues are resolved as desired.  Great project to learn from.  Thanks again for all your time reviewing and providing feedback.

**Comment by thabaum** [comment]
@sbwalker I will provide my updates early tomorrow morning towards the latest "beta.1" version which wont include all the "To Do" list, had to work a bit late tonight... but I want to share my progress on the features and enhancement updates to review prior to my travels when you have time so I can review your feedback when I get time.  Thank you.

**Comment by sbwalker** [comment]
@thabaum after further review of the PR and considerations, these are my preferred recommendations:

- I will not merge this PR in its current form because it essentially solves the same problem as the AutoComplete component and it does not make sense to maintain 2 components which serve the same purpose in the long term.

- I would encourage you to do an analysis between the current AutoComplete component and your Search UI component to determine areas where your component offers additional functionality would would be beneficial for developers (keeping in mind that Oqtane is not focused on providing a full-featured component UI library). Share your analysis so that I can provide feedback before implementing changes.

- modify the existing Pager component to include embedded Search functionality - not as part of a separate component. This will require an additional Parameter for ShowSearch which will be defaulted to "False" so that it is not a breaking change to existing users of the Pager. The Pager search will utilize the data already loaded into the Pager component (a parameter will be needed to specify the SearchColumns being targeted). An optional parameter will provide an OnSearch event hook so that the parent component can implement server-side search logic in some scenarios (this is not not required). The Pager search will have a textbox, Search, and Reset button which will be displayed at the top of the Pager. There will be no "type-ahead" functionality necessary - this will be a simple search based on entering search criteria and clicking the Search button. The search input and buttons should use the input-group class to group UI elements to be consistent with the rest of the Admin UI.

**Comment by thabaum** [comment]
@sbwalker that is great, but at least let me show you my "Beta" version.

The only issue I have is debouncing does not refresh the pager results some odd reason using Timer function.  Task.Delay is ugly results so it would be needed.

This is just to share some final thoughts on what I had accomplished and let you review all the paramters, settings, concerns and maybe we can just add this to the pager.

It maybe a lot like the Autocomplete, but to be honest i have not reviewed that component as it didnt make sense in the naming of it, if it was called "SearchAutoComplete" i would of suggested moving that component into the pager somehow.

But I did a lot of work here, I think you may like what I have done, it does not look much like my "Alpha" version for looking at initial feedback.

I am not worried about any of it being "merged" but the idea of a search in the pager, something simple is great

However this search component can be used in more areas than just the pager, like lets say if you dont use the pager and have a custom table.  I was going to review it for use in other areas possibly tomorrow to show the concept of this UI component at work in more than one area within the Oqtane framework.

But like you said you have two other components so maybe some of this logic can go into one, and some into another... or it was just a fun task to take on and share.

I followed most the big tech guys look and options along with naming conventions for features... you got to take this for a test drive.   I am uploading changes now before heading out so you can review.

I have to say it is bigger than I wanted to do but it covers so many scenarios it is flipping awesome for a basic child component.

I took all your feedback into consideration... again the debouncing is something that can be fine tuned once I find out why the pager does not get its filtered data.

I appreciate your time for feedback regardless if it gets added to the Framework.  There is still fine tuning as I would call this a Beta release with the prior an Alpha.  Take away the debouncing feature desires and this could be close to a release candidate...

It is "false" in the resource file as in "Disabled" so it wont show up on the pager until you modify this setting to "true" to show this all works as you desired for a non-breaking change.

**Comment by thabaum** [comment]
@sbwalker 
I had to re-read your comments as I had minimal amount of time this past week, when it rains it pours type of week.  I was still able to put some time into it Friday night/ Saturday morning and submitted the latest version for you to review which covered a lot of what was discussed before.

I will have to review the embedding directly into the pager logic approach and submit a separate PR for discussions based on what you feel after hearing feedback from this.

Pretty much right after I read the last thread message here I uploaded my latest updates and refactored code... so on.  I put additional comments in the PR for the Search UI Component.

I have kept in mind that Oqtane is not a component library and possibly these all need to be placed into a module that is managed separately and comes preinstalled along with the Admin pages.  

It would be nice to have an "admin dashboard" module template where developers can update the latest current versions as a module package per release (major or minor) so the admin pages or UI controls can be installed/uninstalled from the framework without having to build out Oqtane to remove them once a version is created that would takeover the that part of the application.

I am hoping the latest PR will resolve any concerns and maybe look into why de-bouncing is not working refreshing the data filtered results as expected.  It can be something simple or I am not implementing it correctly which I would love to learn how to... and keep this feature in this UI control.  The `PerformSearch `I believe is firing off it's logic as it lets you know if results are found or not correctly as I added error messaging logic as well to this latest version of the PR.  Something with possibly the results being reset due to search term going empty at some point either in the Parent or Child (Search) component might be the issue or an order of events needs to be different or `bind:event` attribute needs to be used when de-bouncing enabled is another idea I wanted to try and put logic to switch that bind attribute as needed if that's the case.  I was hoping one of the elite programmers here might review and shine some more light on that and we have ourselves one very nice bonus to use anywhere like the pager.

There are some other considerations I mentioned in the PR discussion relating to CSS and some other configs but I am not going to make this as special as some other UI library components, I just was intending on making something KISS but effective for many use cases.

Switches are all in place I used the Regx file like a registry file in Windows more or less to set some settings.  More settings can be adjusted here as well if desired but generally you don't do this, I was making a few exceptions.  And storing in the database makes this UI component more complex but an option.

Again trying to keep these messages here short but give enough insight as to what I was trying to do and if you can read the documentation comments included it should shine further light on each parameter and method.

- if you have less than minimum characters ALL results are shown always
- if you hit cancel it clears all error messages and resets results
- invalid characters message check
- "x" close icon input options
- cancel button options (all but right / left) always on the right, but i can easily set it to the left as an option and include an icon for it if desired instead of text like I did for the `Search` button. I thought it would be a bit overkill...
- CSS dynamics can be introduced as well for moving the input cancel icon/text depending on which configuration and hard code that CSS into the component itself or with a CSS code behind file `Search.razor.css` but this can be set in the app.css for now easy and if someone wraps a DIV around the search component in their parent component they can make their own component CSS as desired depending on the configuration used.
- All parameters could be settings in the resx for extreme control, but I just did what i thought was necessary for general purpose use within the framework.
- Made all parameters "private" so they can be easily managed and updated for defaults.  I am hoping I followed C# coding best practices when naming private/public parameters as well for easy code read-ability by anyone looking to maintain something like default settings or extend it further.

This version is by far more complete, with only the de-bouncing and extensive use of state changes along with any unnecessary code logic introduced while implementing the different features involved.  I removed a lot of what I had and tried to be sure to only leave what was actually needed to be used.  The checks can all be verified as well for the regular expression validation check for example I would like to review that further to ensure it is performed prior to searches from any of the different scenarios from the combinations of parameters that can be utilized.

Off to the races again... just a recap on this components logic.  Sorry to make it so big the documentation shot that file size up to "777" lines of code in the Search UI component!  Scroll down to get to the meat of it... it's not as much to process.  I moved the private parameters to the bottom closer to the logic, but I can move these anywhere as well for code structure readability. 

Again no hurry on my end I would like to fine tune it this week and figure out why de-bouncing does not work.  It is so close to working as expected beautifully I can't not figure out why it is not working for future reference.  I think it is a blazor issue since it is not that great with timer stuff from what I can read and see, or it is my approach, or one simple event happening that should not.  If there is an issue I want to get it to Microsoft to work out prior to .NET 8 release if possible as I have tried cancelation tokens approach a few different ways with same results.  Of coarse this was my last feature I am trying to iron out prior to sending the PR, but I had to show the advancements made covering all other concerns so they can moved forward on.

As you can see this last PR update "beta" so to speak does about everything asked for and then some added concerns of my own.

Look forward to the upcoming developer meeting I have it scheduled in!

Thank you.

**Comment by thabaum** [comment]
@sbwalker After further review of the Autocomplete, Inputlist and Pager components, integration now feels like it should be the best path to follow.  I removed my past comments but left one in the PR.

I think you are correct that this is more a custom component that was with good intent but probably should be on the sidelines from the Oqtane project to limit introducing any extra overhead.

There are a few features that should easily integrate so that this isn't a loss to take ideas from.

I noticed this last PR if search as you type is disabled, and the show button is disabled, it still searches as you type :) which I don't think that was the case in the first PR.  I was going to build some unit tests for this in the end to make sure nothing breaks and results are as expected once I had things to a certain level.

However since integration into existing components sounds like a more viable I will agree with your conclusions.

One place I was hoping to see this feature besides in pages was in logs, messaging and notifications.

So maybe a "search" icon can be added to the end of the pager when it is displayed that enables a search form above or below the pager?  Or enabled from a registry or database setting.

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/d0dcd44c-b8e6-452a-a8a3-f602dae399f8)

Sounds like we have a potential set of potential features list generated and now a final request list to lay out the ground work to be done with respect to the Pager and Autocomplete enhancements for the Admin functionality within Oqtane Framework.

If this is the plan going forward with integrating into Pager and Autocomplete the other PR can be closed. 

I can still with some time to work on this... push out an updated version with only things that are 100% working for all the general use cases.  This last PR introduced a few issues I had resolved in the first PR I need to resolve again, not a big deal like the one I am currently having with de-bouncing I was hoping to figure out for myself regardless just to learn from. I just shared what I have that did most all features I as shooting to include with settings set in different ways as desired.  It does have all the updated changes you had mentioned needed to be addressed.

You are welcome to push some fixes into the existing components now that you have my feedback on them, I am not trying to get in the way of progress here.  At the same time if it helps I can try to attempt a couple PR's on these if you like one for each component (pager/autocomplete) and see if I can come close to getting it right for the Oqtane Framework.  I think you can possibly do it faster than having to explain it to me so it is up to you.

Thanks again for the time in reviewing my PR and work towards resolving this issue.
